addons/point_of_sale/models/pos_order.py

```diff
@@ -117,42 +138,41 @@ def _match_payment_to_invoice(self, order):
     def _process_order(self, pos_order):
         pos_session = self.env['pos.session'].browse(pos_order['pos_session_id'])
         if pos_session.state == 'closing_control' or pos_session.state == 'closed':
+            # basically creates a new pos.session (get is not an appropriate verb)
             pos_order['pos_session_id'] = self._get_valid_session(pos_order).id
         order = self.create(self._order_fields(pos_order))
         prec_acc = order.pricelist_id.currency_id.decimal_places
-        journal_ids = set()
+        PosPayment = self.env['pos.payment']
         for payments in pos_order['statement_ids']:
             if not float_is_zero(payments[2]['amount'], precision_digits=prec_acc):
-                order.add_payment(self._payment_fields(payments[2]))
-            journal_ids.add(payments[2]['journal_id'])
+                payment_dict = payments[2]
+                PosPayment.create(dict(pos_order_id=order.id,
+                                        currency_id=order.pricelist_id.currency_id.id,
+                                        **payment_dict))
 
         if pos_session.sequence_number <= pos_order['sequence_number']:
             pos_session.write({'sequence_number': pos_order['sequence_number'] + 1})
             pos_session.refresh()
 
         if not float_is_zero(pos_order['amount_return'], prec_acc):
-            cash_journal_id = pos_session.cash_journal_id.id
-            if not cash_journal_id:
-                # Select for change one of the cash journals used in this
-                # payment
-                cash_journal = self.env['account.journal'].search([
-                    ('type', '=', 'cash'),
-                    ('id', 'in', list(journal_ids)),
-                ], limit=1)
-                if not cash_journal:
-                    # If none, select for change one of the cash journals of the POS
-                    # This is used for example when a customer pays by credit card
-                    # an amount higher than total amount of the order and gets cash back
-                    cash_journal = [statement.journal_id for statement in pos_session.statement_ids if statement.journal_id.type == 'cash']
-                    if not cash_journal:
-                        raise UserError(_("No cash statement found for this session. Unable to record returned cash."))
-                cash_journal_id = cash_journal[0].id
-            order.add_payment({
+            cash_payment_methods = pos_session\
+                                    .payment_method_ids\
+                                    .filtered(lambda pm: pm.is_cash_count)\
+                                    .sorted(key=lambda pm: pm.id)
```

Comment:
Why sorting them by id?

Reply:
It was a default behavior to choose the first cash payment method for change [return] (if you give money more than what you have to pay). This is to ensure that the first cash payment method in the payment_method_ids is used for the change.